src/toxcore/dht/daemon_state.rs

```diff
@@ -15,25 +13,96 @@ use toxcore::state_format::old::*;
+pub struct DeserializeError{
+    ctx: Context<DeserializeErrorKind>,
+}
+
+impl DeserializeError{
+    /// Return the kind of this error.
+    pub fn kind(&self) -> &DeserializeErrorKind {
+        self.ctx.get_context()
+    }
+
+    pub(crate) fn incomplete(needed: Needed, data: Vec<u8>) -> DeserializeError{
+        DeserializeError::from(DeserializeErrorKind::IncompleteData { needed, data })
+    }
+
+    pub(crate) fn deserialize(error: NomErrorKind, data: Vec<u8>) -> DeserializeError{
+        DeserializeError::from(DeserializeErrorKind::Deserialize { error, data })
+    }
+}
+
+impl Fail for DeserializeError{
+    fn cause(&self) -> Option<&Fail> {
+        self.ctx.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.ctx.backtrace()
+    }
+}
+
+impl fmt::Display for DeserializeError{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.ctx.fmt(f)
+    }
+}
+
+/// The specific kind of error that can occur.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum DeserializeErrorKind {
+    /// Error indicates that object can't be parsed.
     Deserialize {
         /// Parsing error
-        error: ErrorKind,
-        /// DhtState object serialized data
+        error: NomErrorKind,
+        /// object serialized data
         data: Vec<u8>,
     },
-    /// Error indicates that more data is needed to parse serialized DhtState object.
-    #[fail(display = "Bytes of DhtState object should not be incomplete: {:?}, data: {:?}", needed, data)]
+    /// Error indicates that more data is needed to parse serialized object.
     IncompleteData {
         /// Required data size to be parsed
         needed: Needed,
-        /// DhtState object serialized data
+        /// object serialized data
         data: Vec<u8>,
     },
+    /// This enum may grow additional variants, so this makes sure clients
+    /// don't count on exhaustive matching. (Otherwise, adding a new variant
+    /// could break existing code.)
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+
+impl fmt::Display for DeserializeErrorKind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            DeserializeErrorKind::Deserialize { ref error, ref data } => {
+                write!(f, "Deserialize object error: {:?}, data: {:?}", error, data)
+            },
+            DeserializeErrorKind::IncompleteData { ref needed, ref data } => {
+                write!(f, "Bytes of object should not be incomplete: {:?}, data: {:?}", needed, data)
+            },
+            DeserializeErrorKind::__Nonexhaustive => panic!("invalid error"),
+        }
+    }
+}
+
+impl From<DeserializeErrorKind> for DeserializeError{
+    fn from(kind: DeserializeErrorKind) -> DeserializeError{
+        DeserializeError::from(Context::new(kind))
+    }
+}
+
+impl From<Context<DeserializeErrorKind>> for DeserializeError{
+    fn from(ctx: Context<DeserializeErrorKind>) -> DeserializeError{
+        DeserializeError{ ctx }
```

Comment:
Do we use the ctx anywhere?

Reply:
https://github.com/tox-rs/tox/blob/c78f332cfeee955e1ca8121c81eb09dcf0c168d9/src/toxcore/dht/daemon_state.rs#L44

Reply:
Is it a read only property which is not set anywhere?

Reply:
As I know and search, Yes.

Reply:
@kurnevsky ?

Reply:
Also there is formatting issue: `DeserializeError{ ctx }` - no space before `{`.